src/hotspot/share/gc/g1/g1BlockOffsetTable.inline.hpp

```diff
@@ -115,9 +115,8 @@ inline HeapWord* G1BlockOffsetTablePart::block_at_or_preceding(const void* addr)
          "Object crossed region boundary, found offset %u instead of 0",
          (uint) _bot->offset_array(_bot->index_for(_hr->bottom())));
   size_t index = _bot->index_for(addr);
-  // We must make sure that the offset table entry we use is valid.  If
-  // "addr" is past the end, start at the last valid index.
-  index = MIN2(index, _next_offset_index - 1);
+  // We must make sure that the offset table entry we use is valid.
+  assert(index < _next_offset_index, "Precondition");
```

Comment:
I feel a bit uneasy that the caller (`block_start`) allows more than the callee (values `< _hr->end()`), but this is probably the change that is supposed to be coming?
Looks good, but I would have preferred the original change or the upcoming change with some asserts included.

Reply:
I can add the deferred change, instead of splitting it off into a separate PR.

Reply:
> the caller (block_start) allows more than the callee (values < _hr->end())

That path should never be taken, so I am not worried about. I suggested putting that in another PR since those assertions are not obvious from the current ticket title. Anyway, this is subjective.

Reply:
@albertnetymk : I know. Still it seems something is missing to me, but that's subjective. Let's just defer it.